File: src/controller/IMessageHandlers.cpp
/*

  INScore Project

  Copyright (C) 2009,2010  Grame

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

  Grame Research Laboratory, 9 rue du Garet, 69001 Lyon - France
  research@grame.fr

*/

#include "IMessageHandlers.h"
#include "IObject.h"

namespace INScore
{

MsgHandler::msgStatus TSetMethodMsgHandler<Master, void (Master::*)(float)>::operator ()(const IMessage* msg)
{ 
	if ( msg->size() != 1 ) return kBadParameters;
	float val(0);
	if ( !msg->param(0, val) ) return kBadParameters;
	(fObject->*fMethod)( val );
	fSlave->modify();
	return kProcessed;
}


} // end namespoace

File: src/Js/controller/IMessageHandlers.ts
enum msgStatus { 
    kBadAddress,
    kProcessed = 1,
    kProcessedNoChange = 2,
    kBadParameters = 4,
    kCreateFailure = 8,
}

class MsgHandler<C> {
    protected fObject: C;
    protected fMethod: string;    
    static fMsgStatus = msgStatus;
     
    constructor(obj:C, method:string) {
        this.fObject = obj;
        this.fMethod = method;    
    }
    
    executeHandler(msg: IMessage): msgStatus { 
        if ( msg.size() != 1 ) { return msgStatus.kBadParameters; } 
         
        var val: number;
        var castParam: Array<any> = msg.cast_param(0, val); 
	    if ( !castParam['status'] ) { return msgStatus.kBadParameters;}
        val = castParam['val'];
	    
        console.log(this.fObject);
        var method = this.fMethod + ''; 
        this.fObject[method](val);

        console.log(this.fObject);
        	    
        // fSlave->modify();
	    return msgStatus.kProcessed;   
    }    
}

class GetParamMsgHandler<C> {
    protected fObject: C;
    protected fMethod: string;    
     
    constructor(obj:C, method:string) {
        this.fObject = obj;
        this.fMethod = method;    
    }
    
    executeHandler(msg: IMessage): msgStatus { 
        if ( msg.size() != 1 ) { return msgStatus.kBadParameters; } 
         
        var val: number;
        var castParam: Array<any> = msg.cast_param(0, val); 
	    if ( !castParam['status'] ) { return msgStatus.kBadParameters;}
        val = castParam['val'];     
        
        var method = this.fMethod + ''; 
        this.fObject[method](val);

	    return msgStatus.kProcessed;   
    }    
}

//********************************************
class TGetParamMethodHandler<C> extends GetParamMsgHandler<C>  {
    protected fObject: C;
    protected fMethod: string;
    
    constructor(obj: C, method: string) {
        super(obj, method);    
    }
}    
    
class TMethodMsgHandler<C> extends MsgHandler<C> {
    
    constructor(obj: C, method: string) {
        super(obj, method);    
    }
}

class TSetMethodMsgHandler<C> extends MsgHandler<C> {
 
    constructor(obj: C, method: string) {
        super(obj, method);    
    }   
}
